src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,706 @@
+import type NodeGraph from './NodeGraph';
+import type { NodeId, NodeAddress, NodeData, NodeMapping } from './types';
+import type KeyManager from '../keys/KeyManager';
+import type ReverseProxy from '../network/ReverseProxy';
+import type ForwardProxy from '../network/ForwardProxy';
+import type { Host, Hostname, Port } from '../network/types';
+import type { ResourceAcquire } from '../utils';
+import type { Timer } from '../types';
+import Logger from '@matrixai/logger';
+import { StartStop, ready } from '@matrixai/async-init/dist/StartStop';
+import { IdInternal } from '@matrixai/id';
+import NodeConnection from './NodeConnection';
+import * as nodesUtils from './utils';
+import * as nodesErrors from './errors';
+import { utils as validationUtils } from '../validation';
+import * as networkUtils from '../network/utils';
+import GRPCClientAgent from '../agent/GRPCClientAgent';
+import * as agentErrors from '../agent/errors';
+import * as grpcErrors from '../grpc/errors';
+import * as nodesPB from '../proto/js/polykey/v1/nodes/nodes_pb';
+import { RWLock, withF } from '../utils';
+
+type ConnectionAndLock = {
+  connection?: NodeConnection<GRPCClientAgent>;
+  timer?: NodeJS.Timer;
+  lock: RWLock;
+};
+
+/**
+ * Data structure to store all NodeConnections. If a connection to a node n does
+ * not exist, no entry for n will exist in the map. Alternatively, if a
+ * connection is currently being instantiated by some thread, an entry will
+ * exist in the map, but only with the lock (no connection object). Once a
+ * connection is instantiated, the entry in the map is updated to include the
+ * connection object.
+ */
+type NodeConnectionMap = Map<string, ConnectionAndLock>;
+
+/**
+ * This interface is provided to each `NodeConnection` in the connection map
+ * to provide necessary features. The main one being `killSelf` to allow the
+ * `NodeConnection` to tell the `NodeConnectionManager` to destroy the
+ * connection thus removing it from  the connection map.
+ */
+export interface NodeConnectionManagerInterface {
+  withConnF: typeof NodeConnectionManager.prototype.withConnF;
+  getSeedNodes: typeof NodeConnectionManager.prototype.getSeedNodes;
+  sendHolePunchMessage: typeof NodeConnectionManager.prototype.sendHolePunchMessage;
+  killSelf(): Promise<void>;
+}
+
+interface NodeConnectionManager extends StartStop {}
+@StartStop()
+class NodeConnectionManager {
+  protected nodeGraph: NodeGraph;
+  protected keyManager: KeyManager;
+  protected fwdProxy: ForwardProxy;
+  protected revProxy: ReverseProxy;
+  protected logger: Logger;
+  // Reusing terms from network domain:
+  protected connTimeoutTime: number; // - timeout for opening the connection
+  protected connConnectTime: number | undefined; // - TTL for the connection itself
```

Comment:
Why is this potentially `undefined`? Shouldn't we **always** have a TLL for our connections?

Reply:
I use the defaults for both in this case.

Reply:
The keep alive time for network connections is already 20 seconds. But that's if there's no response on the network with the keep alive messages. The 20s was originally derived from the behaviour of UTP I think.

The `connTimeoutTime` here then can be longer therefore, since it's more about when there's no activity on the network connection. It can be 1 minute. Since we don't really bother with holding long-running connections for now reason.

So now we have:

1. `NodeConnectionManager` `connConnectTime`
2. `NodeConnection` `connConnectTime`
3. `GRPCClient` `timeout`

And that's all at the GRPC level.

The low level UTP has a lower default of 20s. I wonder if it makes sense to keep the timeouts the same, or to make the node connection/grpc client timeouts a little longer by default.

It doesn't quite make sense for the node connection timeouts to be less than the network timeouts. So at the very least they are the same, or they are longer.

Reply:
Note to self `ConnectionForward` is `StartStop`, while `NodeConnection` is `CreateDestroy`. There was an attempt to move `ConnectionForward` to `CreateDestroy`, but it was made difficult by some things.

In both cases, the `connConnectTime` is not a property of each class. But instead the property of their respective managers. `ForwardProxy` and `NodeConnectionManager`. Therefore the default should be on the managers first, before on the lower level classes.

Reply:
This is now defaulted.